### DISCUSSION TITLE
"self embedding" Servo in a custom element

### DISCUSSION THREAD

**Discussion by webbeef** [servo/servo#32883] [General]
Over in https://github.com/versotile-org/verso/issues/79 there is some discussion about how to build a Servo based browser with a "chrome" UI written in HTML. The direction that the Verso team is exploring involves writing a compositor / window manager to display and position Servo webviews. I believe that will make it difficult to implement some common UI patterns like overlapping UI element (think about the history autocomplete dropdown that overlays the current tab).

Instead, I proposed to implement with Servo something similar to what Firefox uses for its own UI: the chrome is written in HTML (and still a bit of XUL?) with special APIs/privileges,  and web content is displayed in a `<xul:browser>`  element. This makes it possible to use all the web capabilities when creating the browser UI. 

Is it something that is of any interest for the Servo community? That would look like a `<web-view>` custom element with special properties and events, similar to `<xul:browser>` (implemented here in Gecko: https://searchfox.org/mozilla-central/source/toolkit/content/widgets/browser-custom-element.js).

From an implementation standpoint, I'm not sure what's the best way to approach  that. Can Servo basically embed itself in such a custom element? At which level (GL surface or WebRender display list, or something else) ? Should that be done out of Servo itself (but that doesn't seem possible to add custom elements from the embedder API) ?

Any idea welcome!

**Comment by jdm** [comment]
I would be really interested to discuss changes to the embedder API that would be necessary to prototype this outside of the core Servo engine.

**Reply by webbeef** [reply]
> I would be really interested to discuss changes to the embedder API that would be necessary to prototype this outside of the core Servo engine.

It looks like one prerequisite will be to allow implementing of custom elements out of tree. Is that the kind of changes to the embedder API you are willing to discuss?

**Reply by jdm** [reply]
Yeah, now that we have the crown linter I think it's more feasible for embedders to safely interact with the script layer of Servo than it used to be.

**Comment by webbeef** [comment]
Hi @jdm , I started to play around with defining an html element outside of servo. For that I use:
- https://github.com/webbeef/webviewer which is based on `servoshell` : I added a simple webidl in the `webidls` directory, and trigger the codegen from build.rs
- The `webview-element` branch of my servo fork: https://github.com/webbeef/servo/commit/2bd9ac6b3db2d5c25300bde3a01bed9aac978603 -  I made a change to pass the path to the webidl files to the codegen 

This fails with:

  ```
  File "/home/webbeef/dev/servo/third_party/WebIDL/WebIDL.py", line 468, in finish
      raise WebIDLError(
  WebIDL.WebIDLError: error: Unresolved type '<unresolved scope>::HTMLElement'., /home/webbeef/dev/webviewer/webidls/WebViewElement.webidl line 5:31
  interface HTMLWebViewElement : HTMLElement {
                                 ^

```

The error makes sense since I only pass the webidl the `WebViewElement.webidl` to the codegen. The question is then how to get that to work? Ideally we don't want to rebuild the script crate, so we need to teach codegen to know about some interfaces but not to generate code for all of them. If that works, we still have to find a way to merge the new `InterfaceObjectMapPhf` with the default one. Should that map be optionally passed to Servo by the embedder to override the default one?

**Reply by sagudev** [reply]
This is one of servo's biggest problems, gigantic script crate that is very hard to modularize. I think we firstly need to achieve modularization within script crate, then we can just expose this for embedding. WebIDL complicates stuff even further, but I think finding a way to say to webidl that some types are opaque (are defined somewhere else and are simply available) will be also needed.

I have some rough idea of `script-core` that would provide just bindings utils, then each webidl could be actually separate crate (WebGPU would be interesting candidate due to small amount of external webidl object used). The question is how to performantly (preferably in compile time) register those objects from other crate or even embeddeders (maybe https://github.com/dtolnay/linkme).

**Reply by webbeef** [reply]
@sagudev tell me if I'm wrong but I think that the embedder case is not exactly similar to a modularized Servo script crate. If we look at WebGPU, and manage to move it to its own `script_webgpu` crate, that would look like:
- `script_webgpu` includes WebGPU.webidl
- `script_webgpu` provides the implementation for the binding, depending on the `script_core` crate.
- the "main" `script` crate will depend on `script_webgpu`, so that `Navigator includes NavigatorGPU;` from `Navigator.webidl` does the right thing.
- the webidl codegen needs to learn about these external types and where to get them from. Maybe we can annotate the "includes" with the crate name?

But in the embedder case, we can't modify the main `script` crate anymore to inject new interfaces into other scopes like with internal modularization. Maybe the solution for embedders is to offer a more explicit API to register new properties and/or functions on `window`, `navigator` or in the global JS scope. Lots of other details like the multi-process model to take into account also there... oh my!

**Reply by sagudev** [reply]
In my idea main script create would not depend on script_webgpu (more likely other way around) and the part about using linkme sounds the way to go to me for this. Basically properties of JS objects are done as const arrays of JSPropertySpec (search for [gpu in generated workernavigatorbindings.rs](https://gist.github.com/sagudev/0b0a5f19ef4e6384abcc6babfcbd4875#file-workernavigatorbinding-rs-L791) for example), with linkme we could "add" them in compile (actually link) time, even after main script crate is already compiled. Basically if we can modularize script crate in this way, there would be no real difference between script_webgpu and embeddeders, they would/could use same mechanism (well we would need to mark a lot of stuff in `script*` as `pub`).

The problem happens when there is two way relationship between script modules (they extend each other) ...

**Reply by jdm** [reply]
> @sagudev tell me if I'm wrong but I think that the embedder case is not exactly similar to a modularized Servo script crate. If we look at WebGPU, and manage to move it to its own `script_webgpu` crate, that would look like:
> 
>     * `script_webgpu` includes WebGPU.webidl
> 
>     * `script_webgpu` provides the implementation for the binding, depending on the `script_core` crate.
> 
>     * the "main" `script` crate will depend on `script_webgpu`, so that `Navigator includes NavigatorGPU;` from `Navigator.webidl` does the right thing.
> 
>     * the webidl codegen needs to learn about these external types and where to get them from. Maybe we can annotate the "includes" with the crate name?
> 
> 
> But in the embedder case, we can't modify the main `script` crate anymore to inject new interfaces into other scopes like with internal modularization. Maybe the solution for embedders is to offer a more explicit API to register new properties and/or functions on `window`, `navigator` or in the global JS scope. Lots of other details like the multi-process model to take into account also there... oh my!

Certainly the safest API to offer would be one that exclusively uses channel-based communication with the embedder. It would be slower than allowing direct interaction with the DOM via compiled code, but it would definitely support multiprocess mode.

**Reply by sagudev** [reply]
> Certainly the safest API to offer would be one that exclusively uses channel-based communication with the embedder. It would be slower than allowing direct interaction with the DOM via compiled code, but it would definitely support multiprocess mode.

I think linkme approach would work in multi process mode, but embedder needs to be aware that all code is run in content process (code is injected in compile time in content process). We can still build more safe approach on top of that.

I am just saying that if achieve to modularize script within servo using linkme approach, embeddeder would also be able to benefit. Is this the best/safest way for embeddeders? NO. But I think it is interesting way for modular script.
